sdk/lib/udmihelp/udmihelp.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * PROJECT:     ReactOS User-mode DMI/SMBIOS Helper Functions
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     SMBIOS table parsing functions
+ * COPYRIGHT:   Copyright 2018 Stanislav Motylkov
+ */
+
+#include "precomp.h"
+
+static UINT (WINAPI * pGetSystemFirmwareTable)(DWORD, DWORD, PVOID, DWORD);
+
+/* Load SMBIOS Data */
+PVOID
+LoadSMBiosData(PCHAR * Strings)
+{
+    HANDLE hKernel;
+    PVOID pBuffer = NULL;
+    HKEY hKey;
+    DWORD dwType, dwBytes = 0;
+
+    hKernel = GetModuleHandleW(L"kernel32.dll");
+    if (!hKernel)
+    {
+        return NULL;
+    }
+
+    // try use GetSystemFirmwareTable (works on NT 5.2 and higher)
+    pGetSystemFirmwareTable = (void *)GetProcAddress(hKernel, "GetSystemFirmwareTable");
+    if (pGetSystemFirmwareTable)
+    {
+        dwBytes = pGetSystemFirmwareTable('RSMB', 0, NULL, 0);
+        if (dwBytes > 0)
+        {
+            pBuffer = malloc(dwBytes);
+            if (!pBuffer)
+            {
+                return NULL;
+            }
+            pGetSystemFirmwareTable('RSMB', 0, pBuffer, dwBytes);
+        }
+    }
+    if (dwBytes == 0)
+    {
+        // try use registry (works on NT 5.1)
+        if (RegOpenKeyW(HKEY_LOCAL_MACHINE,
+                        L"SYSTEM\\CurrentControlSet\\Services\\mssmbios\\Data",
+                        &hKey) != ERROR_SUCCESS)
+        {
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             NULL,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        pBuffer = malloc(dwBytes);
+        if (!pBuffer)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             pBuffer,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            free(pBuffer);
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        RegCloseKey(hKey);
+    }
+    ParseSMBiosTables(pBuffer, dwBytes, Strings);
+    return pBuffer;
+}
+
+/* Trim converted DMI string */
+VOID
+TrimDmiStringW(LPWSTR lpStr)
+{
+    unsigned Length;
+    int i = 0;
+
+    if (!lpStr)
+    {
+        return;
+    }
+
+    Length = wcslen(lpStr);
+    if (Length == 0)
+    {
+        return;
+    }
+
+    /* Trim leading spaces */
+    while (i < Length && lpStr[i] <= 32)
+    {
+        i++;
+    }
+    if (i < Length)
+    {
+        if (i > 0)
+        {
+            Length -= i;
+            memcpy(lpStr, &lpStr[i], (Length + 1) * sizeof(WCHAR));
```

Comment:
NO!
`memcpy` has undefined behavior if source and dest overlap.
You need `memmove` here.
Also, the notation `lpStr + i` is more readable than `&lpStr[i]` (at least, in my opinion ;))

Reply:
I've read about difference between `memcpy` and `memmove`, the first one always copies bytes from low to high, and in this situation where we offsetting the string it should be ok, since the source pointer is always greater than destination pointer.

But ok, I've changed it for better safety.